src/interfaces/messages/handlers/messages-get.ts

```diff
@@ -0,0 +1,89 @@
+import type { DataStore } from '../../../store/data-store.js';
+import type { DidResolver } from '../../../did/did-resolver.js';
+import type { MessageStore } from '../../../store/message-store.js';
+import type { MethodHandler } from '../../types.js';
+import type { MessagesGetMessage, MessagesGetReply, MessagesGetReplyEntry } from '../types.js';
+
+import { DataStream } from '../../../utils/data-stream.js';
+import { DwnConstant } from '../../../core/dwn-constant.js';
+import { Encoder } from '../../../utils/encoder.js';
+import { MessageReply } from '../../../core/message-reply.js';
+import { MessagesGet } from '../messages/messages-get.js';
+import { authenticate, authorize } from '../../../core/auth.js';
+import { DwnInterfaceName, DwnMethodName, Message } from '../../../core/message.js';
+
+type HandleArgs = { tenant: string, message: MessagesGetMessage };
+
+export class MessagesGetHandler implements MethodHandler {
+  constructor(private didResolver: DidResolver, private messageStore: MessageStore, private dataStore: DataStore) {}
+
+  public async handle({ tenant, message }: HandleArgs): Promise<MessagesGetReply> {
+    let messagesGet: MessagesGet;
+
+    try {
+      messagesGet = await MessagesGet.parse(message);
+    } catch (e) {
+      return MessageReply.fromError(e, 400);
+    }
+
+    try {
+      await authenticate(message.authorization, this.didResolver);
+      await authorize(tenant, messagesGet);
+    } catch (e) {
+      return MessageReply.fromError(e, 401);
+    }
+
+    const promises: Promise<MessagesGetReplyEntry>[] = [];
+
+    for (const messageCid of message.descriptor.messageCids) {
```

Comment:
should we attempt to do any deduplication here in case the caller provided the same `messageCid` more than once?

Reply:
good idea. done